python/tests/test_writer.py

```diff
@@ -176,6 +176,32 @@ def test_write_modes(tmp_path: pathlib.Path, sample_data: pa.Table):
     assert DeltaTable(path).to_pyarrow_table() == sample_data
 
 
+def test_append_only_should_append_only_with_the_overwrite_mode(
+    tmp_path: pathlib.Path, sample_data: pa.Table
+):
+    path = str(tmp_path)
+
+    config = {"delta.appendOnly": "true"}
+
+    write_deltalake(path, sample_data, mode="append", configuration=config)
+
+    table = DeltaTable(path)
+    write_deltalake(table, sample_data, mode="append")
+
+    fail_modes = ["overwrite", "ignore", "error"]
+
+    for mode in fail_modes:
+        with pytest.raises(
+            ValueError,
+            match="If configuration has delta.appendOnly = 'true', mode must be 'append'",
+        ):
+            write_deltalake(path, sample_data, mode=mode)
+            write_deltalake(table, sample_data, mode=mode)
```

Comment:
If it errors on the first of these, we aren't testing the second, right? Or if it doesn't error on the first line, the test could still pass but we wouldn't like that behavior!

So maybe better to make those separate iterations:

```python
# Test both passing in path string and DeltaTable
possible_args = [path, table]
for arg, mode in itertools.product(possible_args, fail_modes):
    with pytest.raises(...):
        write_deltalake(arg, sample_data, mode=mode)

```

Reply:
Thank you! I was not aware of this feature!

If it had errored on the first we wouldn't have tested the second, but if we hadn't errored on the first we should have tested the second. Either way, I had intended to test both, and your recommended method is great for that.